torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,242 @@
+        )
+
+    def get_mem_consumed(self) -> int:
+        return self.mem_consumed
+
+
+WINFO: Dict[weakref.ref, WeakRefInfo] = WeakIdKeyDictionary()
+
+
+def update_stats():
+    global MEMORY_MAX, WINFO
+    curr_use = 0
+    for k, v in MEMORY_USE.items():
+        curr_use += WINFO[k].get_mem_consumed()
+
+    if MEMORY_MAX < curr_use:
+        MEMORY_MAX = curr_use
+
+
+def track(t):
+    def cb(_):
+        update_stats()
+
+    reftype = RefType.activation
+    if isinstance(t, nn.Parameter):
+        reftype = RefType.parameter
+    st = t.untyped_storage()
+    wt = weakref.ref(st, cb)
+    winfo = WeakRefInfo(st.size(), st.element_size(), reftype)
+    WINFO[st] = winfo
+    MEMORY_USE[st] = wt
+    update_stats()
+
+
+class MemoryTrackingMode(TorchDispatchMode):
+    def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+        res = func(*args, **kwargs or {})
+        tree_map_only(torch.Tensor, track, res)
+        return res
+
+
+def get_current_memory_allocated() -> Dict[str, float]:
+    global MEMORY_USE, WINFO
+    mem_stats = defaultdict(float)
+    mem_stats[RefType.parameter] = 0
+    mem_stats[RefType.gradient] = 0
+    mem_stats[RefType.optstate] = 0
+    mem_stats[RefType.activation] = 0
+    mem_stats[RefType.buffer] = 0
+    for k, v in MEMORY_USE.items():
+        winfo = WINFO[k]
+        mem_stats[winfo.reftype] += winfo.get_mem_consumed()
+    mem_stats["total"] = sum([m for m in mem_stats.values()])
+    return mem_stats
+
+
+def get_fqn() -> str:
+    fqn = ".".join(parents)
+    return fqn
+
+
+def enter_module_forward(name: str):
+    @torch.no_grad
+    def f(module: nn.Module, args: Any):
+        global parents
+        parents.append(name)
+        fqn = get_fqn()
+        memory_tracking[fqn]["Before Forward"] = get_current_memory_allocated()
+
+    return f
+
+
+def exit_module_forward(name: str):
+    @torch.no_grad
+    def f(module: nn.Module, args: Any, outputs: Any):
+        global parents
+        assert parents[-1] == name
+        fqn = get_fqn()
+        memory_tracking[fqn]["After Forward"] = get_current_memory_allocated()
+        parents.pop()
+
+    return f
+
+
+def enter_module_backward(name: str):
+    @torch.no_grad
+    def f(module: nn.Module, grad_output: Any):
+        global parents
+        parents.append(name)
+        fqn = get_fqn()
+        memory_tracking[fqn]["Before Backward"] = get_current_memory_allocated()
+
+    return f
+
+
+def exit_module_backward(name: str):
+    @torch.no_grad
+    def f(module: nn.Module, grad_input: Any, grad_output: Any):
+        global parents
+        assert parents[-1] == name
+        fqn = get_fqn()
+        memory_tracking[fqn]["After Backward"] = get_current_memory_allocated()
+        parents.pop()
+
+    return f
+
+
+def final_call():
+    fqn = get_fqn()
+    memory_tracking[fqn]["After Backward"] = get_current_memory_allocated()
+    parents.pop()
+
+
+def register_hooks(name: str, module: nn.Module):
```

Comment:
We have something similar for flop_counter here
Can we re-use it https://github.com/pytorch/pytorch/blob/d1da9cc654b4560bebb96f2b82923b7cab2e761e/torch/utils/flop_counter.py#L380 ?

Also we have class level hooks at https://pytorch.org/docs/main/generated/torch.nn.modules.module.register_module_forward_hook.html that will trigger on every single Module. This can be useful if you want to trigger the hook on every single module that runs without the user having to give the Module to you to instrument. Or as a way to automatically trigger the instrumentation the first time you see such a Module.

Reply:
Used something similar to flop counter